packages/flutter/lib/src/rendering/box.dart

```diff
@@ -1768,6 +1768,106 @@ abstract class RenderBox extends RenderObject {
+  ///
+  /// Do not override this method. Instead, implement [computeDryLayout].
+  @mustCallSuper
+  Size getDryLayout(BoxConstraints constraints) {
+    bool shouldCache = true;
+    assert(() {
+      // we don't want the checked-mode intrinsic tests to affect
+      // who gets marked dirty, etc.
+      if (RenderObject.debugCheckingIntrinsics)
+        shouldCache = false;
+      return true;
+    }());
+    if (shouldCache) {
+      _cachedDryLayoutSizes ??= <BoxConstraints, Size>{};
+      return _cachedDryLayoutSizes!.putIfAbsent(constraints, () => computeDryLayout(constraints));
+    }
+    return computeDryLayout(constraints);
+  }
+
+  /// Computes the value returned by [getDryLayout]. Do not call this
+  /// function directly, instead, call [getDryLayout].
+  ///
+  /// Override in subclasses that implement [performLayout] or [performResize].
+  /// This method should return the [Size] that this [RenderBox] would like to
+  /// be given the provided [BoxConstraints].
+  ///
+  /// The size returned by this method must match the [size] that the
+  /// [RenderBox] will compute for itself in [performLayout] (or
+  /// [performResize], if [sizedByParent] is true).
+  ///
+  /// If this algorithm depends on the size of a child, the size of that child
+  /// should be obtained using its [getDryLayout] method.
+  ///
+  /// ### When the size cannot be known
+  ///
+  /// There are cases where render objects do not have an efficient way to
+  /// compute their size without doing a full layout. For example, the size
+  /// may depend on the baseline of a child, which is not available without
+  /// doing a full layout.
+  ///
+  /// In such cases, it may be impossible (or at least impractical) to actually
+  /// return a valid answer. In such cases, the function should call
+  /// [debugDryLayoutNotSupported] from within an assert and and return a dummy
+  /// value of `const Size(0, 0)`.
+  @protected
+  Size computeDryLayout(BoxConstraints constraints) {
+    assert(debugDryLayoutNotSupported(
+      'The RenderBox "$runtimeType" does not implement "computeDryLayout".',
+      <DiagnosticsNode>[
+        ErrorHint(
+          'If you are not writing your own RenderBox subclass, then this is not\n'
+          'your fault. Contact support: https://github.com/flutter/flutter/issues/new?template=2_bug.md'
+        ),
+      ],
+    ));
+    return const Size(0, 0);
+  }
+
+  static bool _dryLayoutSupported = true;
+
+  /// Called from [computeDryLayout] within an assert if the given [RenderBox]
+  /// subclass does not support calculating a dry layout.
+  ///
+  /// When asserts are enabled, this method will throw a [FlutterError] with
+  /// the provided `reason` for not supporting dry layout. Otherwise, it will
+  /// simply return true.
+  ///
+  /// See also:
+  ///
+  ///  * [computeDryLayout], which lists some reasons why it may not be feasible
+  ///    to compute the dry layout.
```

Comment:
should this mention debugCheckingIntrinsics? It seems to be the real reason this function exists...